### PR TITLE
Database Schema update

### PR DESCRIPTION
**Pull request by ammanz**
### Changes made/Work done (Instructions on how to test below)
1. Added/removed/moved around a lot of the mock data in the repo (is responsible for >90% of the git diff).
2. Added scripts and relevant instructions to readme on the following:
    1. connecting to locally running mongodb instance
    1. fetching mock data from bandintown and spotify to circumvent rate-limiting and insufficient data
    1. rebuilding devcontainer to apply any new changes to configuration
    1. viewing contents of locally running database
    1. dropping all data from database and running server (`npm run start:clean`)
3. Refactored database schema to the following:
    1. Festivals, Artists, Users, and SpotifyImports are all separate collections. NOTE: matched festivals for user are now computed instead of stored. The Festivals collection stores all of the festivals of all users.
    1. Users store an array of archived and saved festival ids, and an array of "selected artists"
    1. SpotifyImports includes all of the spotify metadata of artists, along with the corresponding user id and time of import
    1. Artists store only relevant information from spotify imports, all listed artists in spotify imports have a reference to their corresponding artists. This allows users to access a centralized list of artists and to see how the artists also evolve if they are listening to the same artists.
4. Refactored the login callback behaviour to the following:
    1. After successful login to Spotify, it checks to see if there was an import in the last 24 hours
    1. If there is one, it skips the rest and redirects user to results page
    1. Otherwise, it does the following:
        - fetches top artists from Spotify and saves the import and any new artists then overwrites the user's current selected artists to all of artists from the import
        - fetches matching festivals from Bandintown and saves them (if festival already exists, it adds any new matching artists references to lineup)
        - redirects user to results page
5. Added/updated the following API endpoints:
    1. a GET matching festivals endpoint: it queries the database for all festivals that have at least one of the user's selected artists
        - added TODO comments on possible performance optimizations and refactoring to allow pagination (needs more research)
    1. a PATCH userlists endpoing in the user router: it updates the saved/archived festivals for the user idempotently

### Setup
1. After pulling changes, run the `clean-dev-container-for-fresh-build.sh` script on Mac terminal, then reopen the vscode window in dev container. This should make some required updates to the local db configuration and also install the mongodb extension automatically that is used to view the contents of the db.
2. Run `npm ci` in both server and react-app folders.
3. Update the server `.env` to connect to the local running database and use mock data if needed (follow readme instructions)

### Login Callback Behaviour
1. Run both frontend and backend on separate terminals in the vscode window. Then click on the spotify login button on the frontend. This should navigate to a blank screen and also fetch and update data in the backend.
2. Navigate to the mongodb vscode extension and connect using the following connection string to the local db: `mongodb://root:password@127.0.0.1:27017/?authSource=admin&readPreference=primary&ssl=false`
3. Inspect and contents of the `festival-fanatic` database. It should be populated with a single spotify import and user, and various festivals and artists. Validate the following:
    1. The user's selectedArtists contains all of the ids to the artists from the spotify import.
    2. The spotifyImport artistRefs point to the artist in the artist collection.
    3. The festivals' artist array point to the matching artists in the artist collection.
4. Navigate back to the login page from the blank page and click on the login button again. It should navigate to the same blank page, but the backend to log out an exception saying that the most recent import was less than 24 hours ago.
5. Close the server process, then comment out lines 143 - 146 in `server/routes/index.js`. This should save an import from spotify every time the user logs in. 
6. Run `npm run start:clean` in the server folder. Once it says db connected, refresh the mongo extension and validate that the database is empty.
7. Login twice consecutively using the frontend, then refresh the mongo extension and validate that no duplicate users/artists/festivals were stored, and that a second (most likely identical) spotify import is stored for the user. The `selectedArtists` should also be updated for the user but since the import is identical, it will seem to be unchanged.

### Updated API Endpoints
1. Start Postman and setup cookie interceptor with google chrome. For domain, use `localhost`
2. Navigate back to login screen and login. Then on Postman, go to the cookies window and select the `festivalFanaticSession` cookie.
3. Open a tab and input `http://localhost:3001/festivals/matched` into the endpoint textfield and select GET. After sending, verify that all of the festivals are present from the database.
4. Duplicate tab and input `http://localhost:3001/users/userlist` into the endpoint textfield and select PATCH.
5. Under request body, select raw json and input the following (use an existing festival id in your running db, mine might not work) then hit send:
```
{
    "festivalId": "62e03c57135ac0f823306bc8",
    "listCategory": "archived",
    "action": "add"
}
```
6. Inspect the user document in the db to validate that it was added to the archive list. Hit send again and validate that a duplicate isn't saved.
7. Update the request body to the following and hit send again:
```
{
    "festivalId": "62e03c57135ac0f823306bc8",
    "listCategory": "archived",
    "action": "remove"
}
```
8. Validate that the festival id is removed from the archived list. Hit send again and validate that no error is thrown.
9. Repeat steps 5 - 8 with ` "listCategory": "saved"` instead and validate that the same behaviour occurs for the saved list.

### CODE REVIEW

**Review by azra-l — APPROVED** [review]
LGTM!

File: server/routes/index.js [hunk 1]
```diff
@@ -105,70 +109,92 @@ router.get("/callback", async function (req, res, next) {
...
+              {
+                userId: hashedId
+              }, {
+              new: true,
+              upsert: true,
+            });
+
+            const currentSpotifyImports = await SpotifyImports.find({ userId: user.userId }).sort({ timeOfImport: -1 }).limit(1);
+            if (currentSpotifyImports.length > 0 && (new Date()).getTime() - currentSpotifyImports[0].timeOfImport.getTime() < 24 * 60 * 60 * 1000) {
+              throw new Error('no need to update imports, last import was done less than a day ago');
```
Comment (StanfordLin) [inline server/routes/index.js:145]: Had to remove the throw otherwise the rest of the stuff will not run, I added to Azra's FE branch

Comment (StanfordLin) [reply server/routes/index.js:145]: LGTM otherwise

Comment (ammanz) [reply server/routes/index.js:145]: Were you connecting to atlas when running this? Or your local db?

If we remove the throw, then we would be importing from spotify and caching the spotify data on every login. We want to only story it once every 24 hours though right?

**Pull request merged**